src/btree/bt_sync_obsolete.c

```diff
@@ -0,0 +1,291 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __sync_obsolete_inmem_evict --
+ *     Check whether the inmem ref is obsolete according to the newest stop time point and mark it
+ *     for urgent eviction.
+ */
+static int
+__sync_obsolete_inmem_evict(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    WT_ADDR_COPY addr;
+    WT_MULTI *multi;
+    WT_PAGE_MODIFY *mod;
+    WT_TIME_AGGREGATE newest_ta;
+    uint32_t i;
+    char tp_string[WT_TP_STRING_SIZE];
+    const char *tag;
+    bool do_visibility_check, obsolete, ovfl_items;
+
+    /*
+     * Skip the modified pages as their reconciliation results are not valid any more. Check for the
+     * page modification only after acquiring the hazard pointer to protect against the page being
+     * freed in parallel.
+     */
+    WT_ASSERT(session, ref->page != NULL);
+    if (__wt_page_is_modified(ref->page))
+        return (0);
+
+    /*
+     * Initialize the time aggregate via the merge initialization, so that stop visibility is copied
+     * across correctly. That is we need the stop timestamp/transaction IDs to start as none,
+     * otherwise we'd never mark anything as obsolete.
+     */
+    WT_TIME_AGGREGATE_INIT_MERGE(&newest_ta);
+    do_visibility_check = obsolete = ovfl_items = false;
+
+    mod = ref->page->modify;
+    if (mod != NULL && mod->rec_result == WT_PM_REC_EMPTY) {
```

Comment:
Can we use a switch block here? I think it would be cleaner.

Reply:
It isn't an obvious change - the sequence of if statements use different pieces of information. OTOH: the decision making in the code isn't as obvious as I'd like it.
https://github.com/wiredtiger/wiredtiger/pull/8589
I pushed another branch that splits this into a sequence of if statements and a switch. Take a look and let me know what you think.